### PR TITLE
Issue #3782 - X-Forwarded-Port refers to node listening port

### PR DESCRIPTION
**Pull request by joakime**
+ Making existing testcases more reliable and easier to maintain.
+ Updating testcases to test requestURL as well.
+ Adding new testcases for `X-Forwarded-Port` where they only modify the port of an existing `Host:` header.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

**Comment by gregw** [comment]
@joakime I like the tests, but not your solution.  I'm working on an alternate solution.

**Comment by gregw** [comment]
@joakime I've pushed an alternate implementation that passes your tests

File: jetty-gcloud/jetty-gcloud-session-manager/pom.xml [hunk 1]
```diff
@@ -33,6 +33,13 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <!-- avoid depending on a range dependency from a transitive dependency -->
+    <dependency>
+      <groupId>io.grpc</groupId>
+      <artifactId>grpc-core</artifactId>
+      <version>1.0.1</version>
```
[the remark this answers is not in the document]
Comment (joakime) [reply jetty-gcloud/jetty-gcloud-session-manager/pom.xml]: You can thank google for that.
They broken their own dependencies.

See: https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-3784/2/pipeline/

[the remark this answers is not in the document]
Comment (joakime) [reply jetty-gcloud/jetty-gcloud-session-manager/pom.xml]: And that change wasn't mine, @olamy pushed a fix for the above CI build issue.
A change that also exists in mainline `jetty-9.4.x`.

File: jetty-server/src/main/java/org/eclipse/jetty/server/ForwardedRequestCustomizer.java [hunk 2]
```diff
@@ -252,6 +252,16 @@ public void setForwardedPortHeader(String forwardedPortHeader)
         }
     }
 
+    public boolean isForwardedPortHeaderRemote()
+    {
+        return _forwardedPortHeaderRemote;
+    }
+
+    public void setForwardedPortHeaderRemote(boolean forwardedPortHeaderRemote)
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/ForwardedRequestCustomizer.java]: Can you document in javadoc what this is?

File: jetty-server/src/main/config/modules/http-forwarded.mod [hunk 3]
```diff
@@ -24,6 +24,7 @@ etc/jetty-http-forwarded.xml
 # jetty.httpConfig.forwardedProtoHeader=X-Forwarded-Proto
 # jetty.httpConfig.forwardedForHeader=X-Forwarded-For
 # jetty.httpConfig.forwardedPortHeader=X-Forwarded-Port
+# jetty.httpConfig.forwardedPortHeaderRemote=false
```
Comment (joakime) [inline jetty-server/src/main/config/modules/http-forwarded.mod]: This probably needs to be documented in this module

**Comment by joakime** [comment]
Darn, I can't approve my own PRs. :-)

File: jetty-gcloud/jetty-gcloud-session-manager/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (joakime) [reply jetty-gcloud/jetty-gcloud-session-manager/pom.xml]: This change, isn't unique for this PR.
Marking as resolved.

**Pull request merged**